src/runtime_src/core/pcie/windows/shim.cpp

```diff
@@ -1105,3 +1147,157 @@ xclRead(xclDeviceHandle handle, enum xclAddressSpace space,
   auto shim = get_shim_object(handle);
   return shim->read(space,offset,hostbuf,size) ? 0 : size;
 }
+
+void queryDeviceWithQR(uint64_t _deviceID, boost::any & _returnValue,
+	int QR_ID, const std::type_info & _typeInfo, uint64_t statClass)
+{
+  xclDeviceHandle handle = xclOpen((int)_deviceID, 0, XCL_INFO);
```

Comment:
A shim level function should work on an already open device handle.   Who ever calls this function should make sure the device is open and pass the handle.  If you are calling this function through a xrt_core::device object, then the device object has already opened the device so all you need is to pass the handle down.